neptune/instance.go

```diff
@@ -2,38 +2,160 @@ package neptune
+
+	if _, err := n.exec(q); err != nil {
+		log.ErrorC("neptune exec failed on AddVersionDetailsToInstance", err, data)
+		return err
+	}
 	return nil
 }
 
+// SetInstanceIsPublished sets a flag on an instance node to indicate the published state
 func (n *NeptuneDB) SetInstanceIsPublished(ctx context.Context, instanceID string) error {
+	data := log.Data{
+		"instance_id": instanceID,
+	}
+
+	q := fmt.Sprintf(query.SetInstanceIsPublished, instanceID)
+
+	if _, err := n.exec(q); err != nil {
+		log.ErrorC("neptune exec failed on SetInstanceIsPublished", err, data)
+		return err
+	}
 	return nil
 }
 
+// CreateInstanceConstraint is not needed for the neptune implementation, as constraints are
+// not a neptune construct
 func (n *NeptuneDB) CreateInstanceConstraint(ctx context.Context, i *model.Instance) error {
 	return nil
 }
 
+// CreateInstance will check if an instance node already exists and create one from
+// the provided details if one does not exist
 func (n *NeptuneDB) CreateInstance(ctx context.Context, i *model.Instance) error {
+	if err := i.Validate(); err != nil {
+		return err
+	}
+
+	data := log.Data{
+		"instance_id": i.InstanceID,
+	}
+
+	exists, err := n.InstanceExists(ctx, i)
+	if err != nil {
+		return err
+	}
+
+	if exists {
+		return nil
+	}
+
+	create := fmt.Sprintf(query.CreateInstance, i.InstanceID, strings.Join(i.CSVHeader, ","))
+	if _, err := n.exec(create); err != nil {
+		log.ErrorC("neptune exec failed on CreateInstance", err, data)
+		return err
+	}
 	return nil
 }
 
+// AddDimensions list to the specified instance node
 func (n *NeptuneDB) AddDimensions(ctx context.Context, i *model.Instance) error {
+	if err := i.Validate(); err != nil {
+		return err
+	}
+
+	data := log.Data{
+		"instance_id": i.InstanceID,
+	}
+
+	q := fmt.Sprintf(query.AddInstanceDimensionsPart, i.InstanceID)
+	for _, d := range i.Dimensions {
+		q += fmt.Sprintf(query.AddInstanceDimensionsPropertyPart, d.(string))
+	}
+
+	if _, err := n.exec(q); err != nil {
+		log.ErrorC("neptune exec failed on AddDimensions", err, data)
+		return err
+	}
+
 	return nil
 }
 
+// CreateCodeRelationship links an instance to a code for the given dimension option
 func (n *NeptuneDB) CreateCodeRelationship(ctx context.Context, i *model.Instance, codeListID, code string) error {
+	if err := i.Validate(); err != nil {
+		return err
+	}
+
+	if len(code) == 0 {
+		return errors.New("code is required but was empty")
+	}
+
+	data := log.Data{
+		"instance_id": i.InstanceID,
+		"code_list":   codeListID,
+		"code":        code,
+	}
+
+	createRelationships := fmt.Sprintf(query.CreateInstanceToCodeRelationship, i.InstanceID, code, codeListID)
+	if res, err := n.exec(createRelationships); err != nil {
+		if len(res) > 0 && res[0].Status.Code == gremgo.StatusScriptEvaluationError &&
+			strings.Contains(res[0].Status.Message, fmt.Sprintf(codeListNotFoundFmt, codeListID)) {
+
+			return errors.Wrapf(err, "code or code list not found", data)
```

Comment:
Same here. I think adding something to convey the context could help make it clearer what it was doing when it failed.